numpy/_core/src/multiarray/mapping.c

```diff
@@ -1963,6 +1963,9 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
     if (tmp_arr && solve_may_share_memory(self, tmp_arr, 1) != 0) {
         Py_SETREF(tmp_arr, (PyArrayObject *)PyArray_NewCopy(tmp_arr, NPY_ANYORDER));
     }
+    if (index_has_memory_overlap(self, index_type, indices, index_num, (PyObject *)tmp_arr) == 1) {
+        Py_SETREF(indices->object, PyArray_NewCopy((PyArrayObject *)indices->object, NPY_KEEPORDER));
```

Comment:
Sorry, I don't know how I missed both pings.  This doesn't work, `indices->` is a C array of index information.

If you add a test like:
```
a[2:][None, a[:-2]] = 3
```
that should just segfault.  The other place where the `index_has_memoverlap` is used actually copies the result (self here and then uses the `Writebackifcopy` mechanism to make that work).

Now, it may be however, that this is sufficient, because it could be that this only is relevant to the fast path (because otherwise the iterator setup can protect us).

---

Either way, not sure that you want to look at it again, which is fine... Just realized that I never responded.